src/binary_descriptors/brief.rs

```diff
@@ -0,0 +1,364 @@
+
+use super::{
+    constants::{BRIEF_PATCH_DIAMETER, BRIEF_PATCH_RADIUS},
+    BinaryDescriptor,
+};
+
+/// BRIEF descriptor as described in [Calonder, et. al. (2010)][calonder].
+///
+/// [calonder]: https://www.cs.ubc.ca/~lowe/525/papers/calonder_eccv10.pdf
+#[derive(Clone, PartialEq)]
+pub struct BriefDescriptor {
+    /// Results of the pairwise pixel intensity tests that comprise this BRIEF
+    /// descriptor.
+    pub bits: Vec<u128>,
+    /// Pixel location and corner score of the keypoint associated with this
+    /// BRIEF descriptor.
+    pub corner: Corner,
+}
+
+impl BinaryDescriptor for BriefDescriptor {
+    fn get_size(&self) -> u32 {
+        (self.bits.len() * 128) as u32
+    }
+    fn compute_hamming_distance(&self, other: &Self) -> u32 {
+        assert_eq!(self.get_size(), other.get_size());
+        self.bits
+            .iter()
+            .zip(other.bits.iter())
+            .fold(0, |acc, x| acc + (x.0 ^ x.1).count_ones())
+    }
+    fn get_bit_subset(&self, bits: &[u32]) -> u128 {
+        assert!(
+            bits.len() <= 128,
+            "Can't extract more than 128 bits (found {})",
+            bits.len()
+        );
+        let mut subset = 0;
+        for b in bits {
+            subset <<= 1;
+            // isolate the bit at index b
+            subset += (self.bits[(b / 128) as usize] >> (b % 128)) % 2;
+        }
+        subset
+    }
+    fn get_position(&self) -> Point<u32> {
+        self.corner.into()
+    }
+}
+
+/// Collection of two points that a BRIEF descriptor uses to generate its bits.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct TestPair {
+    /// The first point in the pair.
+    pub p0: Point<u32>,
+    /// The second point in the pair.
+    pub p1: Point<u32>,
+}
+
+fn local_pixel_average(
+    integral_image: &ImageBuffer<Luma<u32>, Vec<u32>>,
+    x: u32,
+    y: u32,
+    radius: u32,
+) -> u8 {
+    let y_min = if y < radius { 0 } else { y - radius };
+    let x_min = if x < radius { 0 } else { x - radius };
+    let y_max = u32::min(y + radius, integral_image.height() - radius);
+    let x_max = u32::min(x + radius, integral_image.width() - radius);
+
+    let pixel_area = (y_max - y_min + 1) * (x_max - x_min + 1);
+    if pixel_area == 0 {
+        return 0;
+    }
+
+    // UNSAFETY JUSTIFICATION
+    //
+    // Benefit
+    //
+    // Removing the unsafe pixel accesses in this function increases the
+    // runtimes for bench_brief_fixed_test_pairs_1000_keypoints,
+    // bench_brief_random_test_pairs_1000_keypoints, and
+    // bench_rotated_brief_1000_keypoints by about 40%, 30%, and 30%,
+    // respectively.
+    //
+    // Correctness
+    //
+    // The values of x_min, x_max, y_min, and y_max are all bounded between zero
+    // and the integral image dimensions by the checks at the top of this
+    // function.
+    let (bottom_right, top_left, top_right, bottom_left) = unsafe {
+        (
+            integral_image.unsafe_get_pixel(x_max + 1, y_max + 1)[0],
```

Comment:
This can be out of bounds if radius = 1.

Reply:
I don't see the case you've found. Can you walk me through it? I changed some bounds checks here, so maybe this case is covered now?